### ISSUE TITLE
async assertion: show last matcher failure after errors

### ISSUE THREAD

**Issue by pohly** [onsi/gomega#621]
Consider the following fake Kubernetes example:

```
		called := false
		getPod := func(ctx context.Context) (*v1.Pod, error) {
			if !called {
				ginkgo.By("returning fake API error")
				return nil, apierrors.NewTooManyRequests("fake", 10)
			}
			called = true
			pod, err := clientSet.CoreV1().Pods(podNamespace).Get(ctx, podName, metav1.GetOptions{})
			if err != nil {
				return nil, err
			}
			ginkgo.By("returning pod")
			return pod, err
		}
		gomega.Eventually(ctx, framework.HandleRetry(getPod)).WithTimeout(5 * timeout).Should(e2epod.BeInPhase(v1.PodRunning))
```

What happens here is:
- getPod is called, returns pod which is pending
- BeInPhase is called, which returns `false, nil`
- getPod is called again, now always fails
- Eventually times out

It prints:
```
  Timeline >>
  STEP: returning pod @ 01/20/23 09:02:59.116
  STEP: returning fake API error @ 01/20/23 09:02:59.127
  [FAILED] in [It] - /nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/pod/wait_test.go:96 @ 01/20/23 09:03:09.132
  << Timeline

  [FAILED] Timed out after 10.015s.
  Error: function returned error: fake
      <*fmt.wrapError | 0xc0004be520>: {
          msg: "function returned error: fake",
          err: <*errors.StatusError | 0xc0003d80a0>{
              ErrStatus: {
                  TypeMeta: {Kind: "", APIVersion: ""},
                  ListMeta: {
                      SelfLink: "",
                      ResourceVersion: "",
                      Continue: "",
                      RemainingItemCount: nil,
                  },
                  Status: "Failure",
                  Message: "fake",
                  Reason: "TooManyRequests",
                  Details: {
                      Name: "",
                      Group: "",
                      Kind: "",
                      UID: "",
                      Causes: nil,
                      RetryAfterSeconds: 10,
                  },
                  Code: 429,
              },
          },
      }
  In [It] at: /nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/pod/wait_test.go:96 @ 01/20/23 09:03:09.132
```

Note that there is no information about the pod although it was retrieved successfully once. The custom code for polling that I am trying to replace in https://github.com/kubernetes/kubernetes/pull/113298 tries to keep track of API errors and I think it managed to dump some information about what it retrieved earlier. It makes the code fairly complicated, and I don't want to replicate that complexity.

It would be great if Gomega could handle this:
- remember the last value without error from Expect and the last error
- clear the last error each time a new value is stored
- when it times out, dump first the error (current behavior) and then a "previous result" (naming to be decided)

**Comment by pohly** [comment]
I think the second half needs to be:

* If leftErr is not nil (in this case we don't have an actual for this iteration):
   * always emit leftErr and then
   * if a prior iteration yielded a *nil* leftErr then:
     * emit rightErr or matcher.FailureMessage(actual), just as before

So basically I expect two pieces of information:
- the leftErr of the last iteration
- the matcher result of the most recent iteration where there was no leftErr

**Comment by onsi** [comment]
lol, yes my bad.  i've edited my comment.  sounds like we're on the same page.

**Comment by onsi** [comment]
hey @pohly - please give 21f30901ac652d6854e0d5b8043cee4a6331848f a try.  Once I get some feedback and/or a 👍 I'll cut a release.

**Comment by pohly** [comment]
I have created https://github.com/kubernetes/kubernetes/pull/115266 to try out just the gomega update in Kubernetes.

I also updated my on-going work on top of it. That's where I added the unit test from the description. One thing that I noticed is that some error reporting now became simpler:

<details>

```patch
diff --git a/test/e2e/framework/pod/wait_test.go b/test/e2e/framework/pod/wait_test.go
index d1816a2b048..d23b1decf52 100644
--- a/test/e2e/framework/pod/wait_test.go
+++ b/test/e2e/framework/pod/wait_test.go
@@ -129,25 +129,23 @@ In [It] at: wait_test.go:56 <time>
 					Failure: &reporters.JUnitFailure{
 						Type: "failed",
 						Description: `[FAILED] Timed out after <after>.
-Error: function returned error: pods "no-such-pod" not found
-    <*fmt.wrapError>: {
-        msg: "function returned error: pods \"no-such-pod\" not found",
-        err: <framework.transientError>{
-            error: <*errors.StatusError>{
-                ErrStatus: {
-                    TypeMeta: {Kind: "", APIVersion: ""},
-                    ListMeta: {
-                        SelfLink: "",
-                        ResourceVersion: "",
-                        Continue: "",
-                        RemainingItemCount: nil,
-                    },
-                    Status: "Failure",
-                    Message: "pods \"no-such-pod\" not found",
-                    Reason: "NotFound",
-                    Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
-                    Code: 404,
+The function passed to Eventually returned the following error:
+pods "no-such-pod" not found
+    <framework.transientError>: {
+        error: <*errors.StatusError>{
+            ErrStatus: {
+                TypeMeta: {Kind: "", APIVersion: ""},
+                ListMeta: {
+                    SelfLink: "",
+                    ResourceVersion: "",
+                    Continue: "",
+                    RemainingItemCount: nil,
                 },
+                Status: "Failure",
+                Message: "pods \"no-such-pod\" not found",
+                Reason: "NotFound",
+                Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
+                Code: 404,
             },
         },
     }
@@ -156,25 +154,23 @@ In [It] at: wait_test.go:60 <time>
 					},
 					SystemErr: `> Enter [It] not found, retry - wait_test.go:59 <time>
 [FAILED] Timed out after <after>.
-Error: function returned error: pods "no-such-pod" not found
-    <*fmt.wrapError>: {
-        msg: "function returned error: pods \"no-such-pod\" not found",
-        err: <framework.transientError>{
-            error: <*errors.StatusError>{
-                ErrStatus: {
-                    TypeMeta: {Kind: "", APIVersion: ""},
-                    ListMeta: {
-                        SelfLink: "",
-                        ResourceVersion: "",
-                        Continue: "",
-                        RemainingItemCount: nil,
-                    },
-                    Status: "Failure",
-                    Message: "pods \"no-such-pod\" not found",
-                    Reason: "NotFound",
-                    Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
-                    Code: 404,
+The function passed to Eventually returned the following error:
+pods "no-such-pod" not found
+    <framework.transientError>: {
+        error: <*errors.StatusError>{
+            ErrStatus: {
+                TypeMeta: {Kind: "", APIVersion: ""},
+                ListMeta: {
+                    SelfLink: "",
+                    ResourceVersion: "",
+                    Continue: "",
+                    RemainingItemCount: nil,
                 },
+                Status: "Failure",
+                Message: "pods \"no-such-pod\" not found",
+                Reason: "NotFound",
+                Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
+                Code: 404,
             },
         },
     }
@@ -188,25 +184,23 @@ In [It] at: wait_test.go:60 <time>
 					Failure: &reporters.JUnitFailure{
 						Type: "failed",
 						Description: `[FAILED] Timed out after <after>.
-Error: function returned error: pods "no-such-pod" not found
-    <*fmt.wrapError>: {
-        msg: "function returned error: pods \"no-such-pod\" not found",
-        err: <framework.transientError>{
-            error: <*errors.StatusError>{
-                ErrStatus: {
-                    TypeMeta: {Kind: "", APIVersion: ""},
-                    ListMeta: {
-                        SelfLink: "",
-                        ResourceVersion: "",
-                        Continue: "",
-                        RemainingItemCount: nil,
-                    },
-                    Status: "Failure",
-                    Message: "pods \"no-such-pod\" not found",
-                    Reason: "NotFound",
-                    Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
-                    Code: 404,
+The function passed to Eventually returned the following error:
+pods "no-such-pod" not found
+    <framework.transientError>: {
+        error: <*errors.StatusError>{
+            ErrStatus: {
+                TypeMeta: {Kind: "", APIVersion: ""},
+                ListMeta: {
+                    SelfLink: "",
+                    ResourceVersion: "",
+                    Continue: "",
+                    RemainingItemCount: nil,
                 },
+                Status: "Failure",
+                Message: "pods \"no-such-pod\" not found",
+                Reason: "NotFound",
+                Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
+                Code: 404,
             },
         },
     }
@@ -215,25 +209,23 @@ In [It] at: wait_test.go:64 <time>
 					},
 					SystemErr: `> Enter [It] not found, retry with wrappers - wait_test.go:63 <time>
 [FAILED] Timed out after <after>.
-Error: function returned error: pods "no-such-pod" not found
-    <*fmt.wrapError>: {
-        msg: "function returned error: pods \"no-such-pod\" not found",
-        err: <framework.transientError>{
-            error: <*errors.StatusError>{
-                ErrStatus: {
-                    TypeMeta: {Kind: "", APIVersion: ""},
-                    ListMeta: {
-                        SelfLink: "",
-                        ResourceVersion: "",
-                        Continue: "",
-                        RemainingItemCount: nil,
-                    },
-                    Status: "Failure",
-                    Message: "pods \"no-such-pod\" not found",
-                    Reason: "NotFound",
-                    Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
-                    Code: 404,
+The function passed to Eventually returned the following error:
+pods "no-such-pod" not found
+    <framework.transientError>: {
+        error: <*errors.StatusError>{
+            ErrStatus: {
+                TypeMeta: {Kind: "", APIVersion: ""},
+                ListMeta: {
+                    SelfLink: "",
+                    ResourceVersion: "",
+                    Continue: "",
+                    RemainingItemCount: nil,
                 },
+                Status: "Failure",
+                Message: "pods \"no-such-pod\" not found",
+                Reason: "NotFound",
+                Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
+                Code: 404,
             },
         },
     }
@@ -247,25 +239,23 @@ In [It] at: wait_test.go:64 <time>
 					Failure: &reporters.JUnitFailure{
 						Type: "failed",
 						Description: `[FAILED] Timed out after <after>.
-Error: function returned error: pods "no-such-pod" not found
-    <*fmt.wrapError>: {
-        msg: "function returned error: pods \"no-such-pod\" not found",
-        err: <framework.transientError>{
-            error: <*errors.StatusError>{
-                ErrStatus: {
-                    TypeMeta: {Kind: "", APIVersion: ""},
-                    ListMeta: {
-                        SelfLink: "",
-                        ResourceVersion: "",
-                        Continue: "",
-                        RemainingItemCount: nil,
-                    },
-                    Status: "Failure",
-                    Message: "pods \"no-such-pod\" not found",
-                    Reason: "NotFound",
-                    Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
-                    Code: 404,
+The function passed to Eventually returned the following error:
+pods "no-such-pod" not found
+    <framework.transientError>: {
+        error: <*errors.StatusError>{
+            ErrStatus: {
+                TypeMeta: {Kind: "", APIVersion: ""},
+                ListMeta: {
+                    SelfLink: "",
+                    ResourceVersion: "",
+                    Continue: "",
+                    RemainingItemCount: nil,
                 },
+                Status: "Failure",
+                Message: "pods \"no-such-pod\" not found",
+                Reason: "NotFound",
+                Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
+                Code: 404,
             },
         },
     }
@@ -274,25 +264,23 @@ In [It] at: wait_test.go:68 <time>
 					},
 					SystemErr: `> Enter [It] not found, retry with inverted wrappers - wait_test.go:67 <time>
 [FAILED] Timed out after <after>.
-Error: function returned error: pods "no-such-pod" not found
-    <*fmt.wrapError>: {
-        msg: "function returned error: pods \"no-such-pod\" not found",
-        err: <framework.transientError>{
-            error: <*errors.StatusError>{
-                ErrStatus: {
-                    TypeMeta: {Kind: "", APIVersion: ""},
-                    ListMeta: {
-                        SelfLink: "",
-                        ResourceVersion: "",
-                        Continue: "",
-                        RemainingItemCount: nil,
-                    },
-                    Status: "Failure",
-                    Message: "pods \"no-such-pod\" not found",
-                    Reason: "NotFound",
-                    Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
-                    Code: 404,
+The function passed to Eventually returned the following error:
+pods "no-such-pod" not found
+    <framework.transientError>: {
+        error: <*errors.StatusError>{
+            ErrStatus: {
+                TypeMeta: {Kind: "", APIVersion: ""},
+                ListMeta: {
+                    SelfLink: "",
+                    ResourceVersion: "",
+                    Continue: "",
+                    RemainingItemCount: nil,
                 },
+                Status: "Failure",
+                Message: "pods \"no-such-pod\" not found",
+                Reason: "NotFound",
+                Details: {Name: "no-such-pod", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
+                Code: 404,
             },
         },
     }
```
</details>

I supposed `fmt.wrapError function returned error` (= `fmt.Errorf("function returned error: %w")`) was something that Gomega did internally. Getting rid of it looks better.

The test case from the description also works:
```
  [FAILED] Timed out after 10.015s.
  The function passed to Eventually returned the following error:
  fake API error
      <*errors.StatusError | 0xc0009bfa40>: {
          ErrStatus: 
              code: 429
              details:
                retryAfterSeconds: 10
              message: fake API error
              metadata: {}
              reason: TooManyRequests
              status: Failure,
      }
  At one point, however, the function did return successfully.  But Eventually failed because the matcher was not satisfied:
  Expected Pod to be in <v1.PodPhase>: "Running"
  Got instead:
      <*v1.Pod | 0xc000af1b00>: 
          metadata:
            creationTimestamp: null
            name: pending-pod
            namespace: default
          spec:
            containers: null
          status:
            phase: Pending
  In [It] at: /nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/pod/wait_test.go:93 @ 01/23/23 16:44:43.072
```

But it looks like there's a missing line break?

```
  At one point, however, the function did return successfully.
  But Eventually failed because the matcher was not satisfied:
```

**Comment by onsi** [comment]
> I supposed `fmt.wrapError function returned error` (= `fmt.Errorf("function returned error: %w")`) was something that Gomega did internally. Getting rid of it looks better.

Yes that's right.  This turned into a good opportunity to step back and clean up some of the error/failure reporting.

> But it looks like there's a missing line break?

Actually the lack of line-break was intentional.  I continue to rebel against my middle school english teachers telling me I can't start a sentence with "But" ;)

**Comment by thediveo** [comment]
But --> Yet maybe?

**Comment by pohly** [comment]
I'm not against the `But` at the start of the second sentence. The line just becomes very long. If it is meant to be one line, should there really be two spaces instead of one?

**Comment by pohly** [comment]
For comparison:
```
  At one point, however, the function did return successfully.
  But Eventually failed because the matcher was not satisfied:
  Expected Pod to be in <v1.PodPhase>: "Running"
  Got instead:
```

**Comment by onsi** [comment]
I'm perfectly happy adding a new line :)  will update and ship a release soon.

Also - agreed that Yet > But

**Comment by onsi** [comment]
forgot to circle back - this shipped in v1.26.0
